docs/protocols/java_object_graph_spec.md

```diff
@@ -0,0 +1,457 @@
+
+#### Elements header
+
+In most cases, all collection elements are same type and not null, elements header will encode those homogeneous
+information to avoid the cost of writing it for every elements. Specifically, there are four kinds of information
+which will be encoded by elements header, each use one bit:
+
+- If track elements ref, use first bit `0b1` of header to flag it.
+- If collection has null, use second bit `0b10` of header to flag it. If ref tracking is enabled for this
+  element type, this flag is invalid.
+- If collection element types is not declared type, use 3rd bit `0b100` of header to flag it.
+- If collection element types different, use 4rd bit `0b1000` of header to flag it.
+
+By default, all bits are unset, which means all elements won't track ref, all elements are same type,, not null and the
+actual element is the declare type in custom class field.
+
+#### Elements data
+
+Based on the elements header, the serialization of elements data may skip `ref flag`/`null flag`/`element class info`.
+
+`CollectionSerializer#write/read` can be taken as an example.
+
+### Array
+
+#### Primitive array
+
+Primitive array are taken as a binary buffer, serialization will just write the length of array size as an unsigned int,
+then copy the whole buffer into the stream.
+
+Such serialization won't compress the array. If users want to compress primitive array, users need to register custom
+serializers for such types.
+
+#### Object array
+
+Object array is serialized using collection format. Object component type will be taken as collection element generic
+type.
+
+### Map
+
+> All Map serializer must extends `AbstractMapSerializer`.
+
+Format:
+
+```
+| length(unsigned varint) | map header | key value pairs data |
+```
+
+#### Map header
+
+- For `HashMap/LinkedHashMap`, this will be empty.
+- For `TreeMap`, this will be `Comparator`
+- For other `Map`, this may be extra object field info.
+
+#### Map Key-Value data
+
+Map iteration is too expensive, Fury can't compute the header like for collection before since it introduce
+[considerable overhead](https://github.com/alipay/fury/issues/925).
+Users can use `MapFieldInfo` annotation to provide header in advance. Otherwise Fury will use first key-value pair to
+predict header optimistically, and update the chunk header if predict failed at some pair.
+
+Fury will serialize map chunk by chunk, every chunk
+has 127 pairs at most.
+
+```
++----------------+----------------+~~~~~~~~~~~~~~~~~+
+| chunk size: N  |    KV header   |   N*2 objects   |
++----------------+----------------+~~~~~~~~~~~~~~~~~+
+```
+
+KV header:
+
+- If track key ref, use first bit `0b1` of header to flag it.
+- If key has null, use second bit `0b10` of header to flag it. If ref tracking is enabled for this
+  key type, this flag is invalid.
+- If map key type is not declared type, use 3rd bit `0b100` of header to flag it.
+- If map key type different, use 4rd bit `0b1000` of header to flag it.
```

Comment:
how to understand this sentence? does it means the key is different from all other key?

Reply:
Every key has a type, for example, if we have a map of type `Map<Foo, String> map`, and we have a subclass of type `class Bar extends Bar`, and map has value `Map.of(new Foo(), "v1", new Bar(), "v2")`. In such cases, map key will have type `Foo` and `Bar`, Fury will need to write key type for every key instead of write it only once.

Reply:
thanks, now I know that. so actually, the `map key type` should only compare to it's previous sibling key type.

and in some scene, we can use this flag to determine that should we create a new serializer or not.

Reply:
Yep, and if key type of map is different from previous key, this chunk will be finished, and next chunk may choose to assump every key type of map is different if same key types are not too much